feline/src/main/java/com/spotify/feline/FelineTransformer.java

```diff
@@ -65,24 +68,47 @@ static boolean onEnter(
 
       boolean state = FelineRuntime.STATE.get();
 
-      if (!future.isDone() && !state) {
-        FelineRuntime.accept(typeName + "." + methodName + methodSig);
+      if (state) {
+        // Already inside a blocking operation, fast-exit
+        // and keep state as true (i.e. do nothing)
+        return Collections.emptyMap();
+      }
 
-        // Set state to true as to ignore any nested blocked calls, e.g. where one Future delegates
-        // to another.
-        // The state must be set after consumers are invoked above. Consumers can throw exceptions,
-        // in which case we
-        // must not have modified the state first as it would not be reset in onExit().
-        FelineRuntime.STATE.set(true);
+      if (future.isDone()) {
+        // This method call is not actually blocking, so fast-exit
+        // and keep state as false (i.e. do nothing)
+        return Collections.emptyMap();
       }
 
-      return state;
+      final String methodCall = typeName + "." + methodName + methodSig;
+      FelineRuntime.accept(methodCall);
+
+      // Set state to true as to ignore any nested blocked calls, e.g. where one Future delegates
+      // to another.
+      // The state must be set after consumers are invoked above. Consumers can throw exceptions,
+      // in which case we
+      // must not have modified the state first as it would not be reset in onExit().
+      FelineRuntime.STATE.set(true);
+
+      // Can't use custom classes here, since they won't be visible to classes in standard library
+      // such as Future.
```

Comment:
I probably misunderstand how `@Advice.OnMethodEnter` works, but why would the return value need to be visible to other classes? The javadocs for the annotation make it sound like the return value will only be used to pass a value to the method annotated with `OnMethodExit`

Reply:
I had great difficulty trying to add something that would be visible both to consumers of this API as well as internally in the transformer. You could branch off my PR and give it a try if you want, no need to rush this out.

Reply:
ah, I think I see what you mean now - defining a class that would be visible/usable by someone calling `Feline.addOnExitConsumerFirst(Consumer<..>)` as well as within the `@Advice` annotated methods. Gotcha - I was asking more to clarify my understanding of the comment more than suggesting something else should be done.